File: packages/azure_speech/macos/Classes/AzureSpeechPlugin.swift
import Cocoa
import FlutterMacOS
import MicrosoftCognitiveServicesSpeech

public class AzureSpeechPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "azure_speech",
      binaryMessenger: registrar.messenger)
    channel.setMethodCallHandler { call, result in
      guard let args = call.arguments as? [String: Any] else {
        result(FlutterError(code: "bad_args", message: "Expected map", details: nil))
        return
      }
      DispatchQueue.global(qos: .userInitiated).async {
        do {
          let payload: String
          switch call.method {
          case "assess":
            payload = try Self.performAssessment(args: args)
          case "transcribe":
            payload = try Self.performTranscription(args: args)
          case "synthesize":
            payload = try Self.performSynthesis(args: args)
          default:
            DispatchQueue.main.async { result(FlutterMethodNotImplemented) }
            return
          }
          DispatchQueue.main.async { result(payload) }
        } catch let e as NSError where e.domain == "AzureSpeech" {
          let code = e.code == 1 ? "no_speech" : "azure_speech_error"
          DispatchQueue.main.async {
            result(FlutterError(code: code, message: e.localizedDescription, details: nil))
          }
        } catch {
          DispatchQueue.main.async {
            result(FlutterError(code: "azure_speech_error", message: error.localizedDescription, details: nil))
          }
        }
      }
    }
  }

  private static func performAssessment(args: [String: Any]) throws -> String {
    let audioPath = args["audioPath"] as! String
    let referenceText = args["referenceText"] as! String
    let language = args["language"] as! String
    let token = args["token"] as? String
    let subscriptionKey = args["subscriptionKey"] as? String
    let region = args["region"] as! String
    let enableProsody = (args["enableProsody"] as? Bool) ?? true
    let enableMiscue = (args["enableMiscue"] as? Bool) ?? true
    let nbestPhonemeCount = (args["nbestPhonemeCount"] as? Int) ?? 1
    let phonemeAlphabet = (args["phonemeAlphabet"] as? String) ?? "IPA"
    let granularityStr = (args["granularity"] as? String) ?? "Phoneme"

    let granularity: SPXPronunciationAssessmentGranularity
    switch granularityStr {
    case "Word": granularity = .word
    case "FullText": granularity = .fullText
    default: granularity = .phoneme
    }

    let speechConfig: SPXSpeechConfiguration
    if let key = subscriptionKey, !key.isEmpty {
      speechConfig = try SPXSpeechConfiguration(subscription: key, region: region)
    } else {
      speechConfig = try SPXSpeechConfiguration(authorizationToken: token!, region: region)
    }
    speechConfig.speechRecognitionLanguage = language

    guard let audioConfig = SPXAudioConfiguration(wavFileInput: audioPath) else {
      throw NSError(
        domain: "AzureSpeech", code: 4,
        userInfo: [NSLocalizedDescriptionKey: "Could not open audio file"])
    }

    let pronunciationConfig = try SPXPronunciationAssessmentConfiguration(
      referenceText,
      gradingSystem: .hundredMark,
      granularity: granularity,
      enableMiscue: enableMiscue)
    if enableProsody {
      pronunciationConfig.enableProsodyAssessment()
    }
    pronunciationConfig.phonemeAlphabet = phonemeAlphabet
    pronunciationConfig.nbestPhonemeCount = nbestPhonemeCount

    let speechRecognizer = try SPXSpeechRecognizer(
      speechConfiguration: speechConfig, audioConfiguration: audioConfig)

    try pronunciationConfig.apply(to: speechRecognizer)

    let semaphore = DispatchSemaphore(value: 0)
    var jsonOut: String?
    var errOut: NSError?

    try speechRecognizer.recognizeOnceAsync { recognitionResult in
      if recognitionResult.reason == SPXResultReason.recognizedSpeech {
        jsonOut = recognitionResult.properties?.getPropertyBy(
          SPXPropertyId.speechServiceResponseJsonResult)
        if jsonOut == nil || jsonOut!.isEmpty {
          errOut = NSError(
            domain: "AzureSpeech", code: 2,
            userInfo: [NSLocalizedDescriptionKey: "Empty JsonResult"])
        }
      } else if recognitionResult.reason == SPXResultReason.noMatch {
        errOut = NSError(
          domain: "AzureSpeech", code: 1,
          userInfo: [NSLocalizedDescriptionKey: "No speech detected"])
      } else if recognitionResult.reason == SPXResultReason.canceled {
        let cancellationDetails = try? SPXCancellationDetails(
          fromCanceledRecognitionResult: recognitionResult)
        let msg = "\(String(describing: cancellationDetails?.reason)): \(cancellationDetails?.errorDetails ?? "")"
        errOut = NSError(
          domain: "AzureSpeech", code: 3,
          userInfo: [NSLocalizedDescriptionKey: msg])
      } else {
        errOut = NSError(
          domain: "AzureSpeech", code: 3,
          userInfo: [NSLocalizedDescriptionKey: "Unexpected result reason"])
      }
      semaphore.signal()
    }

    semaphore.wait()

    if let e = errOut { throw e }
    guard let json = jsonOut, !json.isEmpty else {
      throw NSError(
        domain: "AzureSpeech", code: 2,
        userInfo: [NSLocalizedDescriptionKey: "No assessment JSON"])
    }
    return json
  }

  private static func performTranscription(args: [String: Any]) throws -> String {
    let audioPath = args["audioPath"] as! String
    let language = args["language"] as! String
    let subscriptionKey = args["subscriptionKey"] as! String
    let region = args["region"] as! String

    let speechConfig = try SPXSpeechConfiguration(subscription: subscriptionKey, region: region)
    speechConfig.speechRecognitionLanguage = language

    guard let audioConfig = SPXAudioConfiguration(wavFileInput: audioPath) else {
      throw NSError(
        domain: "AzureSpeech", code: 4,
        userInfo: [NSLocalizedDescriptionKey: "Could not open audio file"])
    }

    let speechRecognizer = try SPXSpeechRecognizer(
      speechConfiguration: speechConfig, audioConfiguration: audioConfig)

    let semaphore = DispatchSemaphore(value: 0)
    var textOut: String?
    var errOut: NSError?

    try speechRecognizer.recognizeOnceAsync { recognitionResult in
      if recognitionResult.reason == SPXResultReason.recognizedSpeech {
        let text = recognitionResult.text ?? ""
        if text.isEmpty {
          errOut = NSError(
            domain: "AzureSpeech", code: 2,
            userInfo: [NSLocalizedDescriptionKey: "Empty transcription text"])
        } else {
          textOut = text
        }
      } else if recognitionResult.reason == SPXResultReason.noMatch {
        errOut = NSError(
          domain: "AzureSpeech", code: 1,
          userInfo: [NSLocalizedDescriptionKey: "No speech detected"])
      } else if recognitionResult.reason == SPXResultReason.canceled {
        let cancellationDetails = try? SPXCancellationDetails(
          fromCanceledRecognitionResult: recognitionResult)
        let msg = "\(String(describing: cancellationDetails?.reason)): \(cancellationDetails?.errorDetails ?? "")"
        errOut = NSError(
          domain: "AzureSpeech", code: 3,
          userInfo: [NSLocalizedDescriptionKey: msg])
      } else {
        errOut = NSError(
          domain: "AzureSpeech", code: 3,
          userInfo: [NSLocalizedDescriptionKey: "Unexpected result reason"])
      }
      semaphore.signal()
    }

    semaphore.wait()

    if let e = errOut { throw e }
    guard let text = textOut, !text.isEmpty else {
      throw NSError(
        domain: "AzureSpeech", code: 2,
        userInfo: [NSLocalizedDescriptionKey: "No transcription text"])
    }
    return text
  }

  private static func performSynthesis(args: [String: Any]) throws -> String {
    let text = args["text"] as! String
    let language = args["language"] as! String
    let subscriptionKey = args["subscriptionKey"] as! String
    let region = args["region"] as! String
    let voice = args["voice"] as? String

    let speechConfig = try SPXSpeechConfiguration(subscription: subscriptionKey, region: region)
    speechConfig.speechSynthesisLanguage = language
    if let voice = voice, !voice.isEmpty {
      speechConfig.speechSynthesisVoiceName = voice
    }

    let synthesizer = try SPXSpeechSynthesizer(
      speechConfiguration: speechConfig, audioConfiguration: nil)

    let result = try synthesizer.speakText(text)
    if result.reason == SPXResultReason.synthesizingAudioCompleted {
      guard let audio = result.audioData, !audio.isEmpty else {
        throw NSError(
          domain: "AzureSpeech", code: 2,
          userInfo: [NSLocalizedDescriptionKey: "Empty synthesis audio"])
      }
      return audio.base64EncodedString()
    }

    if result.reason == SPXResultReason.canceled {
      let cancellationDetails = try? SPXSpeechSynthesisCancellationDetails(
        fromCanceledSynthesisResult: result)
      let msg = "\(String(describing: cancellationDetails?.reason)): \(cancellationDetails?.errorDetails ?? "")"
      throw NSError(
        domain: "AzureSpeech", code: 3,
        userInfo: [NSLocalizedDescriptionKey: msg])
    }

    throw NSError(
      domain: "AzureSpeech", code: 3,
      userInfo: [NSLocalizedDescriptionKey: "Unexpected synthesis result reason"])
  }
}

File: packages/azure_speech/android/src/main/kotlin/dev/enjoy_player/azure_speech/AzureSpeechPlugin.kt
package dev.enjoy_player.azure_speech

import android.os.Handler
import android.os.Looper
import android.util.Base64
import com.microsoft.cognitiveservices.speech.*
import com.microsoft.cognitiveservices.speech.audio.AudioConfig
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.util.concurrent.Executors

class AzureSpeechPlugin : FlutterPlugin, MethodCallHandler {
  private lateinit var channel: MethodChannel
  private val executor = Executors.newSingleThreadExecutor()
  private val mainHandler = Handler(Looper.getMainLooper())

  override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(binding.binaryMessenger, "azure_speech")
    channel.setMethodCallHandler(this)
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    @Suppress("UNCHECKED_CAST")
    val args = call.arguments as? Map<String, Any?>
    if (args == null) {
      result.error("bad_args", "Expected map arguments", null)
      return
    }
    executor.execute {
      try {
        val payload = when (call.method) {
          "assess" -> assess(args)
          "transcribe" -> transcribe(args)
          "synthesize" -> synthesize(args)
          else -> {
            mainHandler.post { result.notImplemented() }
            return@execute
          }
        }
        mainHandler.post { result.success(payload) }
      } catch (e: Exception) {
        val code = mapErrorCode(e)
        mainHandler.post {
          result.error(code, e.message ?: code, null)
        }
      }
    }
  }

  private fun assess(args: Map<String, Any?>): String {
    val audioPath = args["audioPath"] as String
    val referenceText = args["referenceText"] as String
    val language = args["language"] as String
    val token = args["token"] as? String
    val subscriptionKey = args["subscriptionKey"] as? String
    val region = args["region"] as String
    val enableProsody = args["enableProsody"] as? Boolean ?: true
    val enableMiscue = args["enableMiscue"] as? Boolean ?: true
    val nbestPhonemeCount = (args["nbestPhonemeCount"] as? Number)?.toInt() ?: 1
    val phonemeAlphabet = args["phonemeAlphabet"] as? String ?: "IPA"
    val granularityStr = args["granularity"] as? String ?: "Phoneme"
    val granularity = parseGranularity(granularityStr)

    val config =
      if (!subscriptionKey.isNullOrBlank()) {
        SpeechConfig.fromSubscription(subscriptionKey, region)
      } else {
        SpeechConfig.fromAuthorizationToken(token!!, region)
      }
    try {
      config.speechRecognitionLanguage = language
      val audioConfig = AudioConfig.fromWavFileInput(audioPath)
      try {
        val pronunciationConfig = PronunciationAssessmentConfig(
          referenceText,
          PronunciationAssessmentGradingSystem.HundredMark,
          granularity,
          enableMiscue,
        )
        try {
          if (enableProsody) {
            pronunciationConfig.enableProsodyAssessment()
          }
          pronunciationConfig.setPhonemeAlphabet(phonemeAlphabet)
          pronunciationConfig.setNBestPhonemeCount(nbestPhonemeCount)

          val recognizer = SpeechRecognizer(config, audioConfig)
          try {
            pronunciationConfig.applyTo(recognizer)
            val speechResult = recognizer.recognizeOnceAsync().get()
            when (speechResult.reason) {
              ResultReason.RecognizedSpeech -> {
                val json =
                  speechResult.properties.getProperty(
                    PropertyId.SpeechServiceResponse_JsonResult,
                  )
                if (json.isNullOrBlank()) {
                  throw IllegalStateException("Azure returned empty JsonResult property")
                }
                return json
              }
              ResultReason.NoMatch -> {
                throw NoSpeechException()
              }
              else -> {
                val cancel = CancellationDetails.fromResult(speechResult)
                throw IllegalStateException(
                  "${cancel.reason}: ${cancel.errorDetails}",
                )
              }
            }
          } finally {
            recognizer.close()
          }
        } finally {
          pronunciationConfig.close()
        }
      } finally {
        audioConfig.close()
      }
    } finally {
      config.close()
    }
  }

  private fun transcribe(args: Map<String, Any?>): String {
    val audioPath = args["audioPath"] as String
    val language = args["language"] as String
    val subscriptionKey = args["subscriptionKey"] as String
    val region = args["region"] as String

    val config = SpeechConfig.fromSubscription(subscriptionKey, region)
    try {
      config.speechRecognitionLanguage = language
      val audioConfig = AudioConfig.fromWavFileInput(audioPath)
      try {
        val recognizer = SpeechRecognizer(config, audioConfig)
        try {
          val speechResult = recognizer.recognizeOnceAsync().get()
          when (speechResult.reason) {
            ResultReason.RecognizedSpeech -> {
              val text = speechResult.text
              if (text.isNullOrBlank()) {
                throw IllegalStateException("Azure returned empty transcription text")
              }
              return text
            }
            ResultReason.NoMatch -> {
              throw NoSpeechException()
            }
            else -> {
              val cancel = CancellationDetails.fromResult(speechResult)
              throw IllegalStateException(
                "${cancel.reason}: ${cancel.errorDetails}",
              )
            }
          }
        } finally {
          recognizer.close()
        }
      } finally {
        audioConfig.close()
      }
    } finally {
      config.close()
    }
  }

  private fun synthesize(args: Map<String, Any?>): String {
    val text = args["text"] as String
    val language = args["language"] as String
    val subscriptionKey = args["subscriptionKey"] as String
    val region = args["region"] as String
    val voice = args["voice"] as? String

    val config = SpeechConfig.fromSubscription(subscriptionKey, region)
    try {
      config.speechSynthesisLanguage = language
      if (!voice.isNullOrBlank()) {
        config.speechSynthesisVoiceName = voice
      }
      val synthesizer = SpeechSynthesizer(config)
      try {
        val speechResult = synthesizer.SpeakTextAsync(text).get()
        when (speechResult.reason) {
          ResultReason.SynthesizingAudioCompleted -> {
            val audio = speechResult.audioData
            if (audio == null || audio.isEmpty()) {
              throw IllegalStateException("Azure returned empty synthesis audio")
            }
            return Base64.encodeToString(audio, Base64.NO_WRAP)
          }
          else -> {
            val cancel = SpeechSynthesisCancellationDetails.fromResult(speechResult)
            throw IllegalStateException(
              "${cancel.reason}: ${cancel.errorDetails}",
            )
          }
        }
      } finally {
        synthesizer.close()
      }
    } finally {
      config.close()
    }
  }

  private fun parseGranularity(s: String): PronunciationAssessmentGranularity =
    when (s) {
      "Word" -> PronunciationAssessmentGranularity.Word
      "FullText" -> PronunciationAssessmentGranularity.FullText
      else -> PronunciationAssessmentGranularity.Phoneme
    }

  private fun mapErrorCode(e: Exception): String =
    when (e) {
      is NoSpeechException -> "no_speech"
      else -> "azure_speech_error"
    }

  private class NoSpeechException : RuntimeException("No speech detected in the audio.")
}
